tests/src/test/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManagerSuite.scala

```diff
@@ -0,0 +1,114 @@
+    val context = mock[TaskContext]
+    when(context.taskAttemptId()).thenReturn(taskAttemptId)
+    context
+  }
+
+  test("test basic non-blocking requests") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestNonBlockingLease(context, 100)) { lease =>
+      assert(lease.leaseAmount == 100)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 100)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+  }
+
+  test("test shrinking non-blocking request") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestNonBlockingLease(context, 2048)) { lease =>
+      assert(lease.leaseAmount == 1024)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 1024)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test blocking call that does not block") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestLease(context, 100)) { lease =>
+      assert(lease.leaseAmount == 100)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 100)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test blocking too large required throws") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    assertThrows[IllegalArgumentException] {
+      withResource(GpuMemoryLeaseManager.requestLease(context, 0, 2048)) { _ =>
+      }
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test basic blocking behavior") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val firstContext = mockContext(1)
+    val secondContext = mockContext(2)
+    assert(GpuMemoryLeaseManager.getTotalLease(firstContext) == 0)
+    assert(GpuMemoryLeaseManager.getTotalLease(secondContext) == 0)
+    var childLease: MemoryLease = null
+    val child = new Thread() {
+      override def run(): Unit = {
+        childLease = GpuMemoryLeaseManager.requestLease(secondContext, 900)
+      }
+    }
+    withResource(GpuMemoryLeaseManager.requestLease(firstContext, 900)) { lease =>
+      assert(lease.leaseAmount == 900)
+      child.start()
+      // The child should block so give it some time to come up...
+      Thread.sleep(100)
```

Comment:
Sleeps are almost always an indication of a flaky test, especially a sleep that's this small.  One full-GC hiccup could easily make this essentially a sleep(0) in practice, which means the test doesn't actually test what we want it to, despite that it passes this check.

Reply:
yes, I was being lazy. Thanks for making me fix it.

Reply:
This test is still racy, since the child could fail to block on a lease (or still be in the middle of calculating whether or not it should block) when the main thread is checking for the result of getting a lease.  Completely eliminating the race probably will require exposing the internals of the lease manager to some degree so the test can verify a specific thread has indeed been blocked by the lease manager (e.g.: add a test-only predicate function to the lease manager that can be used to see if a particular thread is known to be blocked by the manager).